### DISCUSSION TITLE
How to Solve the Issue of Continuous Control in a Robotic Hand with MuJoCo

### DISCUSSION THREAD

**Discussion by binbinbai** [google-deepmind/mujoco#1949] [Asking for Help]
Hi,

I'm a student and I'm trying to use MuJoCo for a robotic hand grasping task .

For this robotic hand grasping task that involves a sequence of joint angles **q_tar** with a length of 100, where each frame represents the desired target joint angles (including 3 translational and 3 rotational joints), I encountered an issue during the control process.When I use
```
for i in range(100):
    data.qpos = q_tar[i]
    mujoco.mj_forward(model, data)
```
 the grasping sequence is accurate,the manipulator can grasp object visually. 
 
![accurate](https://github.com/user-attachments/assets/993fd7e4-1371-408e-8968-e3a923d9bf1f)

 However, this approach does not account for collisions or other physical interactions, so the object is not actually grasped. To achieve an actual grasp, I then used 
```
for i in range(100):
    data.ctrl = q_tar[i]
    mujoco.mj_step(model, data)
```
but I found that this control method resulted in significant discrepancies.

![wrong](https://github.com/user-attachments/assets/93057da9-e004-469f-832b-d3dfb06d3b1a)

Here is my actuator configuration:
```
   <actuator>
        <general name="A_ARTx" joint="ARTx" ctrlrange="-2 2" biastype="affine" gainprm="500 0 0" biasprm="0 -500 0" />
        <general name="A_ARTy" joint="ARTy" ctrlrange="-2 2" biastype="affine" gainprm="500 0 0" biasprm="0 -500 0" />
        <general name="A_ARTz" joint="ARTz" ctrlrange="-2 2" biastype="affine" gainprm="500 0 0" biasprm="0 -500 0" />
        <general name="A_ARRx" joint="ARRx" ctrlrange="-3.14 3.14" biastype="affine" gainprm="500 0 0" biasprm="0 -500 0" />
        <general name="A_ARRy" joint="ARRy" ctrlrange="-3.14 3.14" biastype="affine" gainprm="500 0 0" biasprm="0 -500 0" />
        <general name="A_ARRz" joint="ARRz" ctrlrange="-3.14 3.14" biastype="affine" gainprm="500 0 0" biasprm="0 -500 0" />
    </actuator>
```
It's important to note that the joint angle differences between consecutive frames in the q_tar sequence are **not uniform**, with differences ranging from 0.001 to 0.1. I think that when the difference between consecutive frames is very small (0.001), the resulting control signal is also very small, making it difficult to achieve precise control.

How can I solve this issue of continuous control?

If frame-by-frame control is difficult to achieve, is there a feasible method to smoothly move the robotic hand to the joint angles of the last frame, only ensuring that the final frame's joint angles are accurate?

**Comment by Balint-H** [comment]
Hello! Can you explain a bit better the discrepancy you observe between the two cases? The actual hand shape looks similar, it just seems to be offset.

**Reply by binbinbai** [reply]
Hello, thank you for your reply. The problem I want to resolve is the offset shown in the images.The hand shape is fine. @Balint-H
